FlowCrypt/Functionality/Services/Compose Message Service/ComposeMessageService.swift

```diff
@@ -0,0 +1,137 @@
+        atts: [SendableMsg.Attachment]
+    ) -> Result<SendableMsg, ComposeMessageError> {
+        let recipients = contextToSend.recipients
+
+        guard recipients.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        let emails = recipients.map(\.email)
+        let hasContent = emails.filter { $0.hasContent }
+
+        guard emails.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard emails.count == hasContent.count else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard input.isReply || contextToSend.subject?.hasContent ?? false else {
+            return .failure(.validationError(.emptySubject))
+        }
+
+        guard let text = contextToSend.message, text.hasContent else {
+            return .failure(.validationError(.emptyMessage))
+        }
+
+        let subject = input.subjectReplyTitle
+            ?? contextToSend.subject
+            ?? "(no subject)"
+
+        guard let myPubKey = self.dataService.publicKey() else {
+            return .failure(.validationError(.missedPublicKey))
+        }
+
+        switch getPubKeys(for: recipients) {
+        case .success(let allRecipientPubs):
+            let replyToMimeMsg = input.replyToMime
+                .flatMap { String(data: $0, encoding: .utf8) }
+
+            let msg = SendableMsg(
+                text: text,
+                to: recipients.map(\.email),
+                cc: [],
+                bcc: [],
+                from: email,
+                subject: subject,
+                replyToMimeMsg: replyToMimeMsg,
+                atts: atts,
+                pubKeys: allRecipientPubs + [myPubKey]
+            )
+
+            return .success(msg)
+        case .failure(let error):
+            return .failure(.validationError(error))
+        }
+    }
+
+    private func getPubKeys(for recepients: [ComposeMessageRecipient]) -> Result<[String], MessageValidationError> {
+        let pubKeys = recepients.map {
+            ($0.email, contactsService.retrievePubKey(for: $0.email))
+        }
+
+        let emailsWithoutPubKeys = pubKeys.filter { $0.1 == nil }.map(\.0)
+
+        guard emailsWithoutPubKeys.isEmpty else {
+            return .failure(.noPubRecipients(emailsWithoutPubKeys))
+        }
+
+        return .success(pubKeys.compactMap(\.1))
+    }
+
+    // MARK: - Encrypt and Send
+    func encryptAndSend(message: SendableMsg) -> AnyPublisher<Void, ComposeMessageError> {
+        messageGateway.sendMail(mime: encryptMessage(with: message))
+            .mapError { ComposeMessageError.gatewayError($0) }
+            .eraseToAnyPublisher()
+    }
+
+    private func encryptMessage(with msg: SendableMsg) -> Data {
+        do {
+            return try core.composeEmail(msg: msg, fmt: MsgFmt.encryptInline, pubKeys: msg.pubKeys).mimeEncoded
+        } catch {
+            fatalError()
+        }
+    }
```

Comment:
It's possible for this to fail for various reasons. For example, try to run this with an invalid public key. It will produce an error, and this should be handled gratiously.

Reply:
Just "replayed" previous app behaviour. 
Should I fix it in scope of this PR?

Reply:
Got it, let's make another PR then. There, we'll check exactly what happens when we give it some bad input.